presto-native-execution/presto_cpp/main/connectors/arrow_flight/ArrowFlightConnector.cpp

```diff
@@ -0,0 +1,152 @@
+  void AddHeader(const std::string& key, const std::string& value) override {
+    headers.push_back(std::make_pair(key, value));
+  }
+};
+
+FlightDataSource::FlightDataSource(
+    const RowTypePtr& outputType,
+    const std::unordered_map<std::string, std::shared_ptr<ColumnHandle>>&
+        columnHandles,
+    std::shared_ptr<auth::Authenticator> authenticator,
+    memory::MemoryPool* pool,
+    const std::shared_ptr<FlightConfig>& flightConfig)
+    : outputType_{outputType},
+      authenticator_{authenticator},
+      pool_{pool},
+      flightConfig_(flightConfig) {
+  // columnMapping_ contains the real column names in the expected order.
+  // This is later used by projectOutputColumns to filter out unecessary
+  // columns from the fetched chunk.
+  columnMapping_.reserve(outputType_->size());
+
+  for (auto columnName : outputType_->names()) {
+    auto it = columnHandles.find(columnName);
+    VELOX_CHECK(
+        it != columnHandles.end(),
+        "missing columnHandle for column '{}'",
+        columnName);
+
+    auto handle = std::dynamic_pointer_cast<FlightColumnHandle>(it->second);
+    VELOX_CHECK_NOT_NULL(
+        handle,
+        "handle for column '{}' is not an FlightColumnHandle",
+        columnName);
+
+    columnMapping_.push_back(handle->name());
+  }
+}
+
+void FlightDataSource::addSplit(std::shared_ptr<ConnectorSplit> split) {
+  auto flightSplit = std::dynamic_pointer_cast<FlightSplit>(split);
+  VELOX_CHECK(flightSplit, "FlightDataSource recieved wrong type of split");
+
+  auto& locs = flightSplit->locations;
+  Location loc;
+  if (locs.size() > 0) {
+    AFC_ASSIGN_OR_RAISE(loc, Location::Parse(locs[0]));
+  } else {
+    auto defaultHost = flightConfig_->defaultServerHostname();
+    auto defaultPort = flightConfig_->defaultServerPort();
+    VELOX_CHECK(defaultHost.has_value(), "Server Hostname not given");
+    VELOX_CHECK(defaultPort.has_value(), "Server Port not given");
+
+    bool defaultSslEnabled = flightConfig_->defaultServerSslEnabled();
+    AFC_ASSIGN_OR_RAISE(
+        loc,
+        defaultSslEnabled
+            ? Location::ForGrpcTls(defaultHost.value(), defaultPort.value())
+            : Location::ForGrpcTcp(defaultHost.value(), defaultPort.value()));
+  }
+
+  FlightClientOptions clientOpts{
+      .disable_server_verification{!flightConfig_->serverVerify()}};
+  auto certPath = flightConfig_->serverSslCertificate();
+  if (certPath.hasValue()) {
+    std::ifstream file(certPath.value());
+    VELOX_CHECK(file.is_open(), "Could not open TLS certificate");
+    std::string cert(
+        (std::istreambuf_iterator<char>(file)),
+        (std::istreambuf_iterator<char>()));
+    clientOpts.tls_root_certs = cert;
+  }
+
+  AFC_ASSIGN_OR_RAISE(auto client, FlightClient::Connect(loc, clientOpts));
+
+  CallOptionsAddHeaders callOptsAddHeaders{};
+  FlightCallOptions& callOpts = callOptsAddHeaders;
+  AddCallHeaders& headerWriter = callOptsAddHeaders;
+  authenticator_->authenticateClient(
```

Comment:
This is necessary because the URL (`Location`) associated with each Flight Ticket is returned by the initial `GetFlightInfo` call to the Flight Plan Server in the coordinator.

Basically, as per the Arrow Flight standard, each Ticket may need to be redeemed at a different Flight Server, which is not known in advance. Even the list of all possible Flight Server locations is not necessarily known. As such, it is not possible to create a single client and use that to execute all splits.

The simplest solution is to create a new client for every split. There are potentially further optimizations we can do, (for example maintaining a cache of clients with their associated locations, and only creating a new client when a new location is received) but that sort of thing is probably better undertaken in subsequent PRs.